### PR TITLE
Click refactor.

### PR DESCRIPTION
**Pull request by DominikPanic**
:cl: LaKiller8
refactor: Refactors click code.
/:cl:

Refactors click code to a /tg/ standard.

Fixes: #1082 
Fixes: #1120 
Fixes: #1082

### CODE REVIEW

**Comment by spookydonut** [comment]
PR description?

File: code/_onclick/adjacent.dm [hunk 1]
```diff
@@ -123,7 +123,7 @@
 
 // This is necessary for storage items not on your person.
 /obj/item/Adjacent(atom/neighbor)
-	if(neighbor == loc || neighbor == loc.loc) //Item is in the neighbor or something that it holds.
+	if(neighbor == loc || neighbor == loc?.loc) //Item is in the neighbor or something that it holds.
```
Comment (Rohesie) [inline code/_onclick/adjacent.dm]: Which item without `loc` happened to get compared?

Comment (DominikPanic) [reply code/_onclick/adjacent.dm]: *shrug

File: code/_onclick/hud/screen_objects.dm [hunk 2]
```diff
@@ -1,120 +1,308 @@
...
+				var/mob/living/carbon/C = usr
+				C.activate_hand("r")
+
+		if("l_hand")
+			if(iscarbon(usr))
+				var/mob/living/carbon/C = usr
+				C.activate_hand("l")
+
+		if("swap")
+			usr:swap_hand()
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/_onclick/hud/screen_objects.dm]: Yes it damn well is.

File: code/modules/mob/mob_helpers.dm [hunk 3]
```diff
@@ -366,7 +366,7 @@ It's fairly easy to fix if dealing with single letters but not so much with comp
 	return
 
 /mob/proc/is_mob_incapacitated(ignore_restrained)
-	return (stat || stunned || knocked_down || knocked_out || (!ignore_restrained && is_mob_restrained()))
+	return (stat || stunned || knocked_down || knocked_out || istype(loc, /mob/living/carbon/Xenomorph) || (!ignore_restrained && is_mob_restrained()))
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/modules/mob/mob_helpers.dm]: Is it merged yet?

File: code/modules/mob/mob_grab.dm [hunk 4]
```diff
@@ -29,14 +29,20 @@
 		M.stop_pulling()
 	. = ..()
 
+
 /obj/item/grab/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
-	if(user.pulling == user.buckled) return //can't move the thing you're sitting on.
+	if(!istype(user))
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/modules/mob/mob_grab.dm]: Yes

File: code/_onclick/xeno.dm [hunk 5]
```diff
@@ -1,100 +1,77 @@
...
+	selected_ability.use_ability(A)
+
+
+
+/mob/living/carbon/Xenomorph/Boiler/ClickOn(atom/A, params)
+	. = ..()
+	if(istype(A, /obj/screen) || !is_bombarding)
+		return
+	if(isturf(A))
+		bomb_turf(A)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/_onclick/xeno.dm]: Eh, out of scope

File: code/_onclick/xeno.dm [hunk 6]
```diff
@@ -1,100 +1,77 @@
...
-	if(mods["shift"])
-		if(selected_ability && !middle_mouse_toggle)
-			selected_ability.use_ability(A)
-			return 1
+
+/mob/living/carbon/Xenomorph/Queen/CtrlMiddleClickOn(atom/A)
+	. = ..()
+	if(!ovipositor)
+		return
+	if(!isxeno(A) || A == src)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/_onclick/xeno.dm]: Out of scope

File: code/modules/cm_marines/smartgun_mount.dm [hunk 7]
```diff
@@ -53,8 +53,11 @@
 	icon_state = "M56D_gun_e"
 	var/rounds = 0 // How many rounds are in the weapon. This is useful if we break down our guns.
 
-	New()
-		update_icon()
+	
+/obj/item/device/m56d_gun/Initialize()
```
Comment (spookydonut) [inline code/modules/cm_marines/smartgun_mount.dm]: atomise

File: code/game/turfs/closed.dm [hunk 8]
```diff
@@ -135,7 +135,7 @@
 	icon_state = "Intersection"
 
 /turf/closed/attackby(obj/item/W, mob/user)
-
+	. = ..()
```
Comment (spookydonut) [inline code/game/turfs/closed.dm]: atomise

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 9]
```diff
@@ -107,7 +108,8 @@
 	sharp = IS_SHARP_ITEM_SIMPLE
 	edge = 1
 
-/obj/item/tool/wirecutters/New()
+/obj/item/tool/wirecutters/Initialize()
+	. = ..()
```
Comment (spookydonut) [inline code/game/objects/items/tools/maintenance_tools.dm]: atomise

File: code/game/objects/items/storage/storage.dm [hunk 10]
```diff
@@ -9,11 +9,11 @@
 	name = "storage"
 	icon = 'icons/obj/items/storage/storage.dmi'
 	w_class = 3.0
-	var/list/can_hold = new/list() //List of objects which this item can store (if set, it can't store anything else)
-	var/list/cant_hold = new/list() //List of objects which this item can't store (in effect only if can_hold isn't set)
-	var/list/bypass_w_limit = new/list() //a list of objects which this item can store despite not passing the w_class limit
-	var/list/click_border_start = new/list() //In slotless storage, stores areas where clicking will refer to the associated item
-	var/list/click_border_end = new/list()
+	var/list/can_hold = list() //List of objects which this item can store (if set, it can't store anything else)
```
Comment (spookydonut) [inline code/game/objects/items/storage/storage.dm]: atomise

File: code/game/objects/items/devices/coins.dm [hunk 11]
```diff
@@ -12,7 +12,8 @@
 	var/sides = 2
 	var/flags_token = TOKEN_GENERAL
 
-/obj/item/coin/New()
+/obj/item/coin/Initialize()
```
Comment (spookydonut) [inline code/game/objects/items/devices/coins.dm]: atomise

File: code/_onclick/ventcrawl.dm [hunk 12]
```diff
@@ -121,19 +118,19 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
 			client.images += A.pipe_vision_img
 			pipes_shown += A.pipe_vision_img
-	//setMovetype(movement_type | VENTCRAWLING)
+
 
 /mob/living/carbon/Xenomorph/Hunter/add_ventcrawl(obj/machinery/atmospherics/starting_machine)
 	. = ..()
 	cancel_stealth()
 
+
```
Comment (spookydonut) [inline code/_onclick/ventcrawl.dm]: reee whitespace changes

File: code/_onclick/ventcrawl.dm [hunk 13]
```diff
@@ -38,17 +31,20 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
 	if(!is_mob_incapacitated() && pipe)
 		return pipe
 
+
 /mob/living/carbon/monkey/can_ventcrawl()
 	return TRUE
 
+
```
Comment (spookydonut) [inline code/_onclick/ventcrawl.dm]: more conflict bait

File: code/_onclick/ventcrawl.dm [hunk 14]
```diff
@@ -13,20 +15,11 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
-			handle_ventcrawl(A)
-		return TRUE
-
-
 /mob/proc/start_ventcrawl()
 	var/atom/pipe
 	var/list/pipes = list()
 	for(var/obj/machinery/atmospherics/components/unary/U in range(1))
-		if(is_type_in_list(U,GLOB.ventcrawl_machinery) && Adjacent(U))
+		if(is_type_in_list(U, GLOB.ventcrawl_machinery) && Adjacent(U))
```
Comment (spookydonut) [inline code/_onclick/ventcrawl.dm]: atomise

File: code/_onclick/other_mobs.dm [hunk 15]
```diff
@@ -34,9 +15,12 @@
 */
 /mob/living/carbon/monkey/UnarmedAttack(var/atom/A)
 	A.attack_paw(src)
+
```
Comment (spookydonut) [inline code/_onclick/other_mobs.dm]: whitespace

File: code/modules/cm_marines/smartgun_mount.dm (hunk 7 above)
Comment (DominikPanic) [reply code/modules/cm_marines/smartgun_mount.dm]: It literally won't work with this click PR, I am not leaving things broken just so there's a few less files changed. That would be a terrible idea.

File: code/modules/mob/mob_helpers.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rohesie) [reply code/modules/mob/mob_helpers.dm]: The one you've added in `code/modules/mob/living/carbon/xenomorph/life.dm`

File: code/game/objects/items/storage/storage.dm [hunk 16]
```diff
@@ -250,36 +250,39 @@
...
+		return
 
-					I.attack_hand(user)
-					return 1
-	return 0
+	var/obj/item/storage/S = master
+	var/obj/item/I = usr.get_active_held_item()
+	if(I)
+		master.attackby(I, usr)
+
```
Comment (Menshin) [inline code/game/objects/items/storage/storage.dm]:
```suggestion
        return
```

File: code/game/objects/items/storage/storage.dm [hunk 17]
```diff
@@ -250,36 +250,39 @@
...
+	var/list/screen_loc_X = splittext(screen_loc_params[1],":")
+	var/click_x = text2num(screen_loc_X[1]) * 32 + text2num(screen_loc_X[2]) - 144
+
+	for(var/i = 1 to length(S.click_border_start))
+		if(S.click_border_start[i] > click_x || click_x > S.click_border_end[i])
+			continue
+		I = S.contents[i]
+		if(!I)
+			continue
+		if(I == S.contents[length(S.contents)])
```
Comment (Menshin) [inline code/game/objects/items/storage/storage.dm]: ...remove that part.

File: code/_onclick/click.dm [hunk 18]
```diff
@@ -1,145 +1,231 @@
-/*
-	client/Click is called every time a client clicks anywhere, it should never be overridden.
+// 1 decisecond click delay (above and beyond mob/next_move)
+//This is mainly modified by click code, to modify click delays elsewhere, use next_move and changeNext_move()
+/mob/var/next_click	= 0
+
+
+// THESE DO NOT EFFECT THE BASE 1 DECISECOND DELAY OF NEXT_CLICK
+/mob/var/next_move_adjust = 0 //Amount to adjust action/click delays by, + or -
+/mob/var/next_move_modifier = 1 //Value to multiply action/click delays by
```
Comment (Rohesie) [inline code/_onclick/click.dm]: These variables should be defined under `mob_defines.dm`

File: code/_onclick/click.dm [hunk 19]
```diff
@@ -1,145 +1,231 @@
...
 
 	face_atom(A)
 
-	// Special type of click.
-	if(is_mob_restrained())
-		RestrainedClickOn(A)
+	if(next_move > world.time)
+		return
+
+	if(istype(loc, /obj/vehicle/multitile/root/cm_armored))
```
Comment (Rohesie) [inline code/_onclick/click.dm]: Define `click_action()` at the atom level, to return FALSE or null, modify the other definitions to always return TRUE (or invert that logic), and do an `if(loc.click_action(A, src, params)) then return`

File: code/_onclick/click.dm (hunk 18 above)
Comment (DominikPanic) [reply code/_onclick/click.dm]: This is /tg/ though

File: code/_onclick/click.dm (hunk 19 above)
Comment (DominikPanic) [reply code/_onclick/click.dm]: Again, /tg/ code except mech is replaced with a tank :c
But yeah, sure

File: code/_onclick/click.dm [hunk 20]
```diff
@@ -274,50 +437,37 @@
...
+/obj/screen/click_catcher/Click(location, control, params)
+	var/list/modifiers = params2list(params)
+	if(modifiers["middle"] && ishuman(usr))
+		var/mob/living/carbon/human/H = usr
+		H.swap_hand()
+	else
+		var/turf/T = params2turf(modifiers["screen-loc"], get_turf(usr.client ? usr.client.eye : usr), usr.client)
+		params += "&catcher=1"
+		if(T)
+			T.Click(location, control, params)
```
Comment (Rohesie) [inline code/_onclick/click.dm]:
```suggestion
			T?.Click(location, control, params)
```
And remove the `if(T)`

File: code/_onclick/click.dm [hunk 21]
```diff
@@ -274,50 +437,37 @@
...
+	. = 1
```
Comment (Rohesie) [inline code/_onclick/click.dm]:
```suggestion
	return TRUE
```

File: code/_onclick/hud/screen_objects.dm [hunk 22]
```diff
@@ -190,376 +381,226 @@
...
+	. = ..()
+	src.parent = parent
 
-/obj/screen/clicked(var/mob/user)
-	if(!user)	return TRUE
 
-	switch(name)
+/obj/screen/component_button/Click(params)
+	if(parent)
+		parent.component_click(src, params)
```
Comment (Rohesie) [inline code/_onclick/hud/screen_objects.dm]:
```suggestion
		parent?.component_click(src, params)
```
And remove the if.

File: code/_onclick/human.dm [hunk 23]
```diff
@@ -41,10 +41,10 @@
 
 	last_chew = world.time
 
-/mob/living/carbon/human/UnarmedAttack(var/atom/A, var/proximity)
 
+/mob/living/carbon/human/UnarmedAttack(var/atom/A, var/proximity)
```
Comment (Rohesie) [inline code/_onclick/human.dm]:
```suggestion
/mob/living/carbon/human/UnarmedAttack(atom/A, proximity)
```

File: code/_onclick/xeno.dm [hunk 24]
```diff
@@ -1,100 +1,76 @@
...
+	. = ..()
+	if(!selected_ability || middle_mouse_toggle)
+		return
+	selected_ability.use_ability(A)
+
+
+
+/mob/living/carbon/Xenomorph/Boiler/ClickOn(atom/A, params)
+	. = ..()
+	if(istype(A, /obj/screen) || !is_bombarding)
```
Comment (Rohesie) [inline code/_onclick/xeno.dm]: This is being handled by `/obj/screen/Click()` now, no?

File: code/_onclick/xeno.dm [hunk 25]
```diff
@@ -1,100 +1,76 @@
...
+		bomb_turf(A)
+	else
+		bomb_turf(get_turf(A))
+	if(client)
+		client.mouse_pointer_icon = initial(client.mouse_pointer_icon)
+
+
+/mob/living/carbon/Xenomorph/Crusher/ClickOn(atom/A, params)
+	. = ..()
+	if(istype(A, /obj/screen))
```
Comment (Rohesie) [inline code/_onclick/xeno.dm]: Same here, `/obj/screen/Click()` should prevent the click from reaching here, no?

File: code/_onclick/xeno.dm (hunk 24 above)
Comment (DominikPanic) [reply code/_onclick/xeno.dm]: Except it doesn't

File: code/_onclick/xeno.dm (hunk 25 above)
Comment (DominikPanic) [reply code/_onclick/xeno.dm]: Same

File: code/modules/atmospherics/machinery/components/unary_devices/vent_pump.dm [hunk 26]
```diff
@@ -320,6 +320,14 @@
 	pipe_vision_img.plane = ABOVE_HUD_PLANE
 	playsound(loc, 'sound/weapons/bladeslice.ogg', 100, 1)
 
+
+/obj/machinery/atmospherics/components/unary/vent_pump/AltClick(mob/living/user)
```
Comment (Rohesie) [inline code/modules/atmospherics/machinery/components/unary_devices/vent_pump.dm]:
```suggestion
/obj/machinery/atmospherics/components/unary/vent_pump/AltClick(mob/user)
```
And define `can_ventcrawl()` as a mob proc, else this has the risk of causing runtimes.

Comment (Rohesie) [reply code/modules/atmospherics/machinery/components/unary_devices/vent_pump.dm]: Actually scratch that. `handle_ventcrawl()` already checks for `can_ventcrawl()`. Redefine that one to the mob level.

File: code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm [hunk 27]
```diff
@@ -331,6 +331,13 @@
 	pipe_vision_img.plane = ABOVE_HUD_PLANE
 	playsound(loc, 'sound/weapons/bladeslice.ogg', 100, 1)
 
+
+/obj/machinery/atmospherics/components/unary/vent_scrubber/AltClick(mob/living/user)
+	if(!user.can_ventcrawl())
```
Comment (Rohesie) [inline code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm]: Same here.

File: code/modules/vehicles/tank.dm [hunk 28]
```diff
@@ -319,4 +319,29 @@
...
 		log_combat(gunner, null, "[src] took [damage] [type] damage [ismob(attacker) ? "from [key_name(attacker)]" : ""].")
 	if(driver)
-		log_combat(driver, null, "[src] took [damage] [type] damage [ismob(attacker) ? "from [key_name(attacker)]" : ""].")
\ No newline at end of file
+		log_combat(driver, null, "[src] took [damage] [type] damage [ismob(attacker) ? "from [key_name(attacker)]" : ""].")
+
+
+/obj/vehicle/multitile/root/cm_armored/proc/click_action(A, mob/user, params)
+	var/list/mods = params2list(params)
+	if(istype(A, /obj/screen) || A == src || mods["middle"] || mods["shift"] || mods["alt"])
```
Comment (Rohesie) [inline code/modules/vehicles/tank.dm]: This should as well be handled by the screen Click redefinition, no?

Comment (DominikPanic) [reply code/modules/vehicles/tank.dm]: As I said relic /tg/ code handles mechs in the same way. Larger refactor needed

**Pull request merged**